go-controller/pkg/node/port_claim.go

```diff
@@ -120,61 +145,61 @@ func updateServicePortClaim(oldSvc, newSvc *kapi.Service) []error {
-		if err != nil {
-			socketError = err
-			break
-		}
-		socket = listener
+		localPort, portError = utilnet.NewLocalPort(desc, ip, "", int(port), utilnet.TCP)
 	case kapi.ProtocolUDP:
-		addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf(":%d", port))
-		if err != nil {
-			socketError = err
-			break
-		}
-		conn, err := net.ListenUDP("udp", addr)
-		if err != nil {
-			socketError = err
-			break
-		}
-		socket = conn
+		localPort, portError = utilnet.NewLocalPort(desc, ip, "", int(port), utilnet.UDP)
 	case kapi.ProtocolSCTP:
 		// Do not open ports for SCTP, ref: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md#the-solution-in-the-kubernetes-sctp-support-implementation
 		return nil
 	default:
-		socketError = fmt.Errorf("unknown protocol %q", protocol)
+		portError = fmt.Errorf("unknown protocol %q", protocol)
 	}
-	if socketError != nil {
-		p.emitPortClaimEvent(svc, port, socketError)
-		return socketError
+	if portError != nil {
+		p.emitPortClaimEvent(svc, port, portError)
+		return portError
 	}
 	p.activeSocketsLock.Lock()
 	defer p.activeSocketsLock.Unlock()
-	if _, exists := p.activeSockets[protocol]; exists {
-		p.activeSockets[protocol][port] = socket
+
+	if _, exists := p.portsMap[*localPort]; exists {
+		return fmt.Errorf("error try to open socket for svc: %s/%s on port: %v again", svc.Namespace, svc.Name, port)
 	} else {
-		p.activeSockets[protocol] = map[int32]activeSocket{
-			port: socket,
+		closeable, err := utilnet.ListenPortOpener.OpenLocalPort(localPort)
+		if err != nil {
+			p.emitPortClaimEvent(svc, port, err)
+			return err
 		}
+		p.portsMap[*localPort] = closeable
 	}
 	return nil
 }
 
-func (p *portClaimWatcher) close(port int32, protocol kapi.Protocol, svc *kapi.Service) error {
+func (p *portClaimWatcher) close(desc string, ip string, port int32, protocol kapi.Protocol, svc *kapi.Service) error {
 	p.activeSocketsLock.Lock()
 	defer p.activeSocketsLock.Unlock()
-	klog.V(5).Infof("Closing socket claimed for service: %s/%s and port: %v", svc.Namespace, svc.Name, port)
-	if socket, exists := p.activeSockets[protocol][port]; exists {
-		if err := socket.Close(); err != nil {
+
+	var localPort *utilnet.LocalPort
+	var err error
+	if protocol == kapi.ProtocolTCP {
+		localPort, err = utilnet.NewLocalPort(desc, ip, "", int(port), utilnet.TCP)
+	} else if protocol == kapi.ProtocolUDP {
+		localPort, err = utilnet.NewLocalPort(desc, ip, "", int(port), utilnet.TCP)
+	}
+	if err != nil {
+		return fmt.Errorf("error localPort creation for svc: %s/%s on port: %v, err: %v", svc.Namespace, svc.Name, port, err)
+	}
+	if _, exists := p.portsMap[*localPort]; exists {
+		if err = p.portsMap[*localPort].Close(); err != nil {
 			return fmt.Errorf("error closing socket for svc: %s/%s on port: %v, err: %v", svc.Namespace, svc.Name, port, err)
 		}
-		delete(p.activeSockets[protocol], port)
+		p.portsMap[*localPort] = nil
```

Comment:
why not delete it? This is a memory leak as written.

Reply:
Oops, you're right, fixed.